client/memory.cpp

```diff
@@ -199,6 +240,34 @@ clEnqueueReadBufferRect(cl_command_queue command_queue, cl_mem buffer, cl_bool b
 	}
 }
 
+SO_EXPORT CL_API_ENTRY cl_int CL_API_CALL
+clEnqueueUnmapMemObject(cl_command_queue command_queue, cl_mem memobj, void* mapped_ptr,
+                        cl_uint num_events_in_wait_list, const cl_event* event_wait_list,
+                        cl_event* event) CL_API_SUFFIX__VERSION_1_0
+{
+	auto conn = gConnection.get();
+	auto buf = conn.getBuffer(GetID(memobj));
+
+	if (buf == nullptr) return CL_INVALID_VALUE;
+	if (buf->data.get() != mapped_ptr) return CL_INVALID_VALUE;
+
+	cl_int ret = CL_SUCCESS;
+
+	if (buf->flags & CL_MAP_WRITE) {
+		ret = clEnqueueWriteBuffer(command_queue, memobj, true, buf->offset,
+                                   buf->size, buf->data.get(),
+                                   num_events_in_wait_list, event_wait_list, event);
+	}
+
+	conn.unregisterBuffer(GetID(memobj));
+	if (ret != CL_SUCCESS) {
+		clReleaseMemObject(memobj);
+		return ret;
+	}
+
+	return clReleaseMemObject(memobj);
```

Comment:
Does this look right to you?
I'm not sure what the sequence here should be.
Right now I'm releasing the mem object in all cases, even if the clEnqueueWriteBuffer() call failed.